internal/component/loki/source/file/scheduler.go

```diff
@@ -0,0 +1,138 @@
+package file
+
+import (
+	"context"
+	"iter"
+	"sync"
+
+	"github.com/grafana/dskit/backoff"
+)
+
+// Scheduler is used to manage sources.
+// So it is up to the users to ensure proper synchronization.
+type Scheduler[K comparable] struct {
+	ctx     context.Context
+	cancel  context.CancelFunc
+	sources map[K]scheduledSource[K]
+
+	running sync.WaitGroup
+}
+
+func NewScheduler[K comparable]() *Scheduler[K] {
+	ctx, cancel := context.WithCancel(context.Background())
+
+	return &Scheduler[K]{
+		ctx:     ctx,
+		cancel:  cancel,
+		sources: make(map[K]scheduledSource[K]),
+	}
+}
+
+// ApplySource will register and start run provided source.
+// If a source with the same key already exists it will do nothing.
+func (s *Scheduler[K]) ApplySource(source Source[K]) {
+	k := source.Key()
+	if _, ok := s.sources[k]; ok {
+		return
+	}
+
+	ctx, cancel := context.WithCancel(s.ctx)
+	st := scheduledSource[K]{
+		ctx:    ctx,
+		cancel: cancel,
+		source: source,
+	}
+
+	s.sources[k] = st
+
+	s.running.Go(func() {
```

Comment:
Nice! new API :D

Reply:
Yeah, just a convenient api added in go 1.25